### PR TITLE
Implement proof of validation

### PR DESCRIPTION
**Pull request by pablomendezroyo**
Implement proof of validation.

Depends on web3signer implementation

### CODE REVIEW

File: packages/brain/src/modules/cron/sendProofOfAttestation.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+): Promise<void> {
+  try {
+    // Get the proofs of attestation from the signer
+    const proofsOfAttestations = await getProofsOfAttestations(
+      signerApi,
+      brainDb,
+      network
+    );
+    if(proofsOfAttestations.length === 0) return
+    await dappnodeSigningProoverApi.sendProofOfAttestation(
```
Comment (Marketen) [inline packages/brain/src/modules/cron/sendProofOfAttestation.ts:28]: this `await` could waste some seconds if our server doesnt respond quickly

File: packages/brain/src/modules/apiClients/dappnodeSignerProover.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        baseUrl: validatorsMonitorUrl || "https://dappnode-signatures.io",
+      },
+      network
+    );
+  }
+
+  public async sendProofOfAttestation(
+    proofOfAttestations: DappnodeSigningProoverPostRequest[]
+  ): Promise<void> {
+    await this.request({
```
Comment (Marketen) [inline packages/brain/src/modules/apiClients/dappnodeSignerProover.ts:23]: I'd define a timeout here to ensure we do not wait too many seconds in case our server is not answering when being sent the attestations

File: packages/brain/src/modules/apiClients/dappnodeSignerProover.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  constructor(network: Network, validatorsMonitorUrl?: string) {
+    super(
+      {
+        baseUrl: validatorsMonitorUrl || "https://dappnode-signatures.io",
+      },
+      network
+    );
+  }
+
+  public async sendProofOfAttestation(
```
Comment (Marketen) [inline packages/brain/src/modules/apiClients/dappnodeSignerProover.ts:20]: I would rename to `sendProofsOfAttestationBatch` (or something similar) since we're sending `proofOfAttestations: DappnodeSigningProoverPostRequest[]` and not a single attestation

File: packages/brain/src/modules/apiClients/web3signer.ts [hunk 4]
```diff
@@ -50,14 +62,37 @@ export class Web3SignerApi extends StandardApi {
         method: "POST",
         endpoint: path.join(this.signEndpoint, pubkey),
         body: JSON.stringify(signerVoluntaryExitRequest),
-        setOrigin: true,
+        headers: this.originHeader,
```
Comment (Marketen) [inline packages/brain/src/modules/apiClients/web3signer.ts:65]: why is this changed in this PR?

File: packages/brain/src/modules/cron/sendProofOfAttestation.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  const signerDappnodeSignRequest: Web3signerPostSignDappnodeRequest = {
+    type: "PROOF_OF_VALIDATION",
+    platform: "dappnode",
+    timestamp: new Date().toISOString(),
+  };
+  // get pubkeys detauls from db
+  const dbPubkeysDetails = brainDb.getData();
+  // For each pubkey, get the proof of attestation from the signer
+  const proofsOfAttestations = await Promise.all(
+    Object.keys(dbPubkeysDetails).map(async (pubkey) => {
```
Comment (Marketen) [inline packages/brain/src/modules/cron/sendProofOfAttestation.ts:54]: maybe it is a good idea to wait a bit between each request to the signer to ensure we dont overload it. just a thought

**Review by Marketen — CHANGES_REQUESTED** [review]

File: packages/brain/src/modules/cron/sendProofOfAttestation.ts (hunk 5 above)
Comment (pablomendezroyo) [reply packages/brain/src/modules/cron/sendProofOfAttestation.ts:50]: already tested in the web3signer with hundred of keys, seems to be handled without issues

File: packages/brain/src/index.ts [hunk 6]
```diff
@@ -7,6 +7,7 @@ import {
   Beaconchain,
   BeaconchaApi,
   ValidatorApi,
+  DappnodeSigningProover,
```
Comment (dappnodedev) [inline packages/brain/src/index.ts:10]: I  I think it`s "Prover", but I'd call it Signature Verifier

File: packages/brain/src/index.ts [hunk 7]
```diff
@@ -69,6 +77,10 @@ export const beaconchainApi = new Beaconchain(
   { baseUrl: beaconchainUrl },
   network
 );
+export const dappnodeSignerProoverApi = new DappnodeSigningProover(
```
Comment (dappnodedev) [inline packages/brain/src/index.ts:80]: Same as above

File: packages/brain/src/index.ts [hunk 8]
```diff
@@ -88,19 +100,26 @@ await brainDb.initialize(signerApi, validatorApi);
...
 
 // CRON
-export const cron = new Cron(
+export const reloadValidatorsCron = new CronJob(
   60 * 1000,
-  signerApi,
-  signerUrl,
-  validatorApi,
-  brainDb
+  new ReloadValidators(
```
Comment (dappnodedev) [inline packages/brain/src/index.ts:105]: We saw that this was causing issues. I would export a function instead of passing here a public method of a  new object

File: packages/brain/src/index.ts [hunk 9]
```diff
@@ -88,19 +100,26 @@ await brainDb.initialize(signerApi, validatorApi);
...
+    signerApi,
+    signerUrl,
+    validatorApi,
+    brainDb
+  ).reloadValidators
+);
+reloadValidatorsCron.start();
+const proofOfAttestationCron = new CronJob(shareCronInterval, () =>
```
Comment (dappnodedev) [inline packages/brain/src/index.ts:113]: Why here attestation and below validation? I would be consistent with the naming

File: packages/brain/src/modules/apiClients/dappnodeSignerProover.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import { StandardApi } from "./index.js";
+import path from "path";
+import {
+  Network,
+  DappnodeSigningProoverPostRequest,
+} from "@stakingbrain/common";
+
+export class DappnodeSigningProover extends StandardApi {
```
Comment (dappnodedev) [inline packages/brain/src/modules/apiClients/dappnodeSignerProover.ts:8]: Here you call it `SigningProover` but the file name is different

File: packages/brain/src/modules/apiClients/dappnodeSignerProover.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  private dappnodeSignEndpoint = "/signatures";
+
+  constructor(network: Network, validatorsMonitorUrl?: string) {
+    super(
+      {
+        baseUrl: validatorsMonitorUrl || "https://dappnode-signatures.io",
```
Comment (dappnodedev) [inline packages/brain/src/modules/apiClients/dappnodeSignerProover.ts:14]: Maybe set this in params like `defaultValidatorsMonitorUrl`

File: packages/brain/src/modules/apiClients/standard.ts [hunk 12]
```diff
@@ -72,10 +74,26 @@ export class StandardApi {
       req = https.request(this.requestOptions);
     } else req = http.request(this.requestOptions);
 
-    if (setOrigin)
```
Comment (dappnodedev) [inline packages/brain/src/modules/apiClients/standard.ts:75]: Why do you remove this? Are you sure it's not necessary?

File: packages/brain/src/modules/apiClients/web3signer.ts [hunk 13]
```diff
@@ -50,14 +68,41 @@ export class Web3SignerApi extends StandardApi {
...
+   */
+  public async signDappnodeProofOfValidation({
+    signerDappnodeSignRequest,
+    pubkey,
+  }: {
+    signerDappnodeSignRequest: Web3signerPostSignDappnodeRequest;
+    pubkey: string;
+  }): Promise<Web3signerPostSignDappnodeResponse> {
+    try {
+      return await this.request({
```
Comment (dappnodedev) [inline packages/brain/src/modules/apiClients/web3signer.ts:90]: Do you think we need a retry system for these requests?

File: packages/brain/src/modules/cron/cron.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+export class CronJob {
+  private task: NodeJS.Timeout | null = null;
+  private interval: number;
+
+  constructor(interval: number, private jobFunction: () => Promise<void>) {
+    this.interval = interval;
+  }
+
+  start(): void {
```
Comment (dappnodedev) [inline packages/brain/src/modules/cron/cron.ts:12]: I would specify here `public` for all these functions

File: packages/brain/src/modules/cron/reloadValidators.ts
Comment (dappnodedev) [inline packages/brain/src/modules/cron/reloadValidators.ts:1]: I assume this has been moved, but not modified

File: packages/common/src/types/api/web3signer/types.ts [hunk 15]
```diff
@@ -56,4 +65,22 @@ export interface Web3SignerPostSignvoluntaryexitRequest {
...
+  platform: "dappnode";
+  timestamp: string;
+}
+
+export interface Web3signerPostSignDappnodeResponse {
+  signature: string;
+  payload: string;
+}
+
+export interface DappnodeSigningProoverPostRequest {
```
Comment (dappnodedev) [inline packages/common/src/types/api/web3signer/types.ts:81]: Same as in the comments above, it's Prover

File: packages/brain/src/modules/envs/index.ts [hunk 16]
```diff
@@ -69,6 +72,15 @@ export function loadStakerConfig(): {
       )}`
     );
 
+  const shareDataWithDappnode = process.env.SHARE_DATA_WITH_DAPPNODE === "true";
+  const validatorsMonitorUrl = process.env.VALIDATORS_MONITOR_URL || "";
+  const shareCronInterval = parseInt(process.env.SHARE_CRON_INTERVAL || "");
+  if (shareDataWithDappnode && (!validatorsMonitorUrl || !shareCronInterval)) {
+    throw Error(
+      "SHARE_DATA_WITH_DAPPNODE is true, but VALIDATORS_MONITOR_URL or SHARE_CRON_INTERVAL are not set"
```
Comment (dappnodedev) [inline packages/brain/src/modules/envs/index.ts:80]: Will this make the brain restart?

File: packages/brain/src/modules/apiClients/web3signer.ts (hunk 13 above)
Comment (Marketen) [reply packages/brain/src/modules/apiClients/web3signer.ts:90]: I would be careful when asking web3signer to try again and again for signing keys, might affect attestation performance

File: packages/brain/src/modules/cron/cron.ts [hunk 17]
```diff
@@ -0,0 +1,34 @@
+import { setInterval, clearInterval } from "timers";
+import logger from "../logger/index.js";
+
+export class CronJob {
+  private task: NodeJS.Timeout | null = null;
+  private interval: number;
+
+  constructor(interval: number, private jobFunction: () => Promise<void>) {
+    this.interval = interval;
```
Comment (Marketen) [inline packages/brain/src/modules/cron/cron.ts:9]: what if interval is negative or 0? will the cronjob execute nonstop? Id put something to prevent it like
```
if (interval <= 0) {
      throw new Error("Interval must be a positive number.");
    }
```

File: packages/brain/src/modules/cron/sendProofsOfValidation.ts [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+  brainDb: BrainDataBase,
+  dappnodeSigningProoverApi: DappnodeSigningProover
+): Promise<void> {
+  try {
+    // Get the proofs of attestation from the signer
+    const proofsOfAttestations = await getProofsOfValidation(
+      signerApi,
+      brainDb
+    );
+    if (proofsOfAttestations.length === 0) return;
```
Comment (Marketen) [inline packages/brain/src/modules/cron/sendProofsOfValidation.ts:24]: maybe log something if `proofsOfAttestations.length === 0` before returning?

**Review by Marketen — APPROVED** [review]

File: packages/brain/src/modules/apiClients/web3signer.ts (hunk 13 above)
Comment (Marketen) [reply packages/brain/src/modules/apiClients/web3signer.ts:90]: I would be careful when asking web3signer to try again and again for signing keys, might affect attestation performance

File: packages/brain/src/modules/envs/index.ts (hunk 16 above)
Comment (pablomendezroyo) [reply packages/brain/src/modules/envs/index.ts]: fixed!

File: packages/brain/src/modules/cron/sendProofsOfValidation.ts [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+    timestamp: Date.now().toString(),
+  };
+  // get pubkeys detauls from db
+  const dbPubkeysDetails = brainDb.getData();
+
+  // only send proof of validation if the user has enabled it
+  // or if there is a stader pubkey
+  if (
+    !shareDataWithDappnode &&
+    !Object.values(dbPubkeysDetails).some((pubkey) => pubkey.tag === "stader")
```
Comment (dappnodedev) [inline packages/brain/src/modules/cron/sendProofsOfValidation.ts:61]: But here if there is a stader pubkey will the brain send the validation proofs for all the pubkeys even if shareDataWithDappnode is disabled?

**Review by dappnodedev — APPROVED** [review]
